File: dataforge-core/src/main/java/hep/dataforge/values/ValueUtils.java
package hep.dataforge.values;

import java.io.IOException;
import java.io.ObjectInput;
import java.io.ObjectOutput;
import java.io.Serializable;
import java.math.BigDecimal;
import java.math.BigInteger;
import java.time.Instant;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;


/**
 * Created by darksnake on 06-Aug-16.
 */
public class ValueUtils {

    public static final Comparator<Number> NUMBER_COMPARATOR = new NumberComparator();
    public static final Comparator<Value> VALUE_COMPARATPR = new ValueComparator();

    public static int compare(Value val1, Value val2) {
        switch (val1.valueType()) {
            case NUMBER:
                return NUMBER_COMPARATOR.compare(val1.numberValue(), val2.numberValue());
            case BOOLEAN:
                return Boolean.compare(val1.booleanValue(), val2.booleanValue());
            case STRING:
                //use alphanumeric comparator here
                return val1.stringValue().compareTo(val2.stringValue());
            case TIME:
                return val1.timeValue().compareTo(val2.timeValue());
            case NULL:
                return val2.valueType() == ValueType.NULL ? 0 : -1;
            default:
                throw new RuntimeException("Uncompareable value");
        }
    }

    /**
     * Checks if given value is between {@code val1} and {@code val1}. Could
     * throw ValueConversionException if value conversion is not possible.
     *
     * @param val1
     * @param val2
     * @return
     */
    public static boolean isBetween(Value val, Value val1, Value val2) {
        return (compare(val, val1) > 0 && compare(val, val2) < 0)
                || (compare(val, val2) > 0 && compare(val, val1) < 0);

    }

    public static boolean isBetween(Object val, Value val1, Value val2) {
        return isBetween(Value.of(val), val1, val2);
    }


    /**
     * Fast and compact serialization for values
     *
     * @param oos
     * @param value
     * @throws IOException
     */
    public static void writeValue(ObjectOutput oos, Value value) throws IOException {
        if (value.isList()) {
            oos.writeChar('L'); // List designation
            oos.writeShort(value.listValue().size());
            for (Value subValue : value.listValue()) {
                writeValue(oos, subValue);
            }
        } else {
            switch (value.valueType()) {
                case NULL:
                    oos.writeChar('0'); // null
                    break;
                case TIME:
                    oos.writeChar('T');//Instant
                    oos.writeLong(value.timeValue().getEpochSecond());
                    oos.writeLong(value.timeValue().getNano());
                    break;
                case STRING:
                    //TODO add encding specification
                    oos.writeChar('S');//String
                    oos.writeUTF(value.stringValue());
                    break;
                case NUMBER:
                    Number num = value.numberValue();
                    if (num instanceof Double) {
                        oos.writeChar('D'); // double
                        oos.writeDouble(num.doubleValue());
                    } else if (num instanceof Integer) {
                        oos.writeChar('I'); // integer
                        oos.writeInt(num.intValue());
                    } else if (num instanceof BigDecimal) {
                        oos.writeChar('B'); // BigDecimal
                        byte[] bigInt = ((BigDecimal) num).unscaledValue().toByteArray();
                        int scale = ((BigDecimal) num).scale();
                        oos.writeShort(bigInt.length);
                        oos.write(bigInt);
                        oos.writeInt(scale);
                    } else {
                        oos.writeChar('N'); //custom number
                        oos.writeObject(num);
                    }
                    break;
                case BOOLEAN:
                    if (value.booleanValue()) {
                        oos.writeChar('+'); //true
                    } else {
                        oos.writeChar('-'); // false
                    }
                    break;
                default:
                    oos.writeChar('C');//custom
                    oos.writeObject(value);
            }
        }
    }

    /**
     * Value deserialization
     *
     * @param ois
     * @return
     * @throws IOException
     * @throws ClassNotFoundException
     */
    public static Value readValue(ObjectInput ois) throws IOException, ClassNotFoundException {
        char c = ois.readChar();
        switch (c) {
            case 'L':
                short listSize = ois.readShort();
                List<Value> valueList = new ArrayList<>();
                for (int i = 0; i < listSize; i++) {
                    valueList.add(readValue(ois));
                }
                return Value.of(valueList);
            case '0':
                return Value.NULL;
            case 'T':
                Instant time = Instant.ofEpochSecond(ois.readLong(), ois.readLong());
                return Value.of(time);
            case 'S':
                return Value.of(ois.readUTF());
            case 'D':
                return Value.of(ois.readDouble());
            case 'I':
                return Value.of(ois.readInt());
            case 'B':
                short intSize = ois.readShort();
                byte[] intBytes = new byte[intSize];
                ois.read(intBytes);
                int scale = ois.readInt();
                BigDecimal bdc = new BigDecimal(new BigInteger(intBytes), scale);
                return Value.of(bdc);
            case 'N':
                return Value.of(ois.readObject());
            case '+':
                return BooleanValue.TRUE;
            case '-':
                return BooleanValue.FALSE;
            case 'C':
                return (Value) ois.readObject();
            default:
                throw new RuntimeException("Wrong value serialization format. Designation " + c + " is unexpected");
        }
    }

    private static class ValueComparator implements Comparator<Value>, Serializable {
        @Override
        public int compare(Value o1, Value o2) {
            return ValueUtils.compare(o1, o2);
        }
    }

    private static class NumberComparator implements Comparator<Number>, Serializable {
        private static final double RELATIVE_NUMERIC_PRECISION = 1e-5;

        private static boolean isSpecial(final Number x) {
            boolean specialDouble = x instanceof Double
                    && (Double.isNaN((Double) x) || Double.isInfinite((Double) x));
            boolean specialFloat = x instanceof Float
                    && (Float.isNaN((Float) x) || Float.isInfinite((Float) x));
            return specialDouble || specialFloat;
        }

        private static BigDecimal toBigDecimal(final Number number) {
            if (number instanceof BigDecimal) {
                return (BigDecimal) number;
            }
            if (number instanceof BigInteger) {
                return new BigDecimal((BigInteger) number);
            }
            if (number instanceof Byte || number instanceof Short
                    || number instanceof Integer || number instanceof Long) {
                return new BigDecimal(number.longValue());
            }
            if (number instanceof Float || number instanceof Double) {
                return new BigDecimal(number.doubleValue());
            }

            try {
                return new BigDecimal(number.toString());
            } catch (final NumberFormatException e) {
                throw new RuntimeException("The given number (\"" + number
                        + "\" of class " + number.getClass().getName()
                        + ") does not have a parsable string representation", e);
            }
        }

        @Override
        public int compare(final Number x, final Number y) {
            double d1 = x.doubleValue();
            double d2 = y.doubleValue();
            if ((d1 != 0 || d2 != 0) && (Math.abs(d1 - d2) / Math.max(d1, d2) < RELATIVE_NUMERIC_PRECISION)) {
                return 0;
            } else if (isSpecial(x) || isSpecial(y)) {
                return Double.compare(d1, d2);
            } else {
                return toBigDecimal(x).compareTo(toBigDecimal(y));
            }
        }
    }

}

File: dataforge-core/src/main/kotlin/hep/dataforge/values/ValueUtils.kt
/*
 * Copyright  2018 Alexander Nozik.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

package hep.dataforge.values

import hep.dataforge.io.IOUtils
import hep.dataforge.providers.Path
import hep.dataforge.providers.Provider
import java.io.IOException
import java.io.ObjectInput
import java.io.ObjectOutput
import java.io.Serializable
import java.math.BigDecimal
import java.math.BigInteger
import java.nio.ByteBuffer
import java.time.Instant
import java.util.*

data class ValueRange(override val start: Value, override val endInclusive: Value) : ClosedRange<Value> {
    operator fun contains(any: Any): Boolean {
        return contains(Value.of(any))
    }
}

operator fun Value.rangeTo(other: Value): ValueRange = ValueRange(this, other)


/**
 * Created by darksnake on 06-Aug-16.
 */
object ValueUtils {

    val NUMBER_COMPARATOR: Comparator<Number> = NumberComparator()

    /**
     * Fast and compact serialization for values
     *
     * @param oos
     * @param value
     * @throws IOException
     */
    @Throws(IOException::class)
    @JvmStatic
    fun writeValue(oos: ObjectOutput, value: Value) {
        if (value.isList) {
            oos.write('*'.toInt()) // List designation
            oos.writeShort(value.list.size)
            for (subValue in value.list) {
                writeValue(oos, subValue)
            }
        } else {
            when (value.type) {
                ValueType.NULL -> oos.writeChar('0'.toInt()) // null
                ValueType.TIME -> {
                    oos.write('T'.toInt())//Instant
                    oos.writeLong(value.time.epochSecond)
                    oos.writeLong(value.time.nano.toLong())
                }
                ValueType.STRING -> {
                    oos.writeChar('S'.toInt())//String
                    IOUtils.writeString(oos, value.string)
                }
                ValueType.NUMBER -> {
                    val num = value.number
                    when (num) {
                        is Double -> {
                            oos.write('D'.toInt()) // double
                            oos.writeDouble(num.toDouble())
                        }
                        is Int -> {
                            oos.write('I'.toInt()) // integer
                            oos.writeInt(num.toInt())
                        }
                        is Long -> {
                            oos.write('L'.toInt())
                            oos.writeLong(num.toLong())
                        }
                        is BigDecimal -> {
                            oos.write('B'.toInt()) // BigDecimal
                            val bigInt = num.unscaledValue().toByteArray()
                            val scale = num.scale()
                            oos.writeShort(bigInt.size)
                            oos.write(bigInt)
                            oos.writeInt(scale)
                        }
                        else -> {
                            oos.write('N'.toInt()) //custom number
                            oos.writeObject(num)
                        }
                    }
                }
                ValueType.BOOLEAN -> if (value.boolean) {
                    oos.write('+'.toInt()) //true
                } else {
                    oos.write('-'.toInt()) // false
                }
                ValueType.BINARY -> {
                    val binary = value.binary
                    oos.write('X'.toInt())
                    oos.write(binary.limit())
                    oos.write(binary.array())
                }
//                else -> {
//                    oos.write('C'.toInt())//custom
//                    oos.writeObject(value)
//                }
            }
        }
    }

    /**
     * Value deserialization
     *
     * @param ois
     * @return
     * @throws IOException
     * @throws ClassNotFoundException
     */
    @Throws(IOException::class, ClassNotFoundException::class)
    @JvmStatic
    fun readValue(ois: ObjectInput): Value {
        return readValue(ois, ois.read().toChar())
    }

    /**
     * Read value knowing its type-character
     */
    fun readValue(ois: ObjectInput, type: Char): Value {
        when (type) {
            '*' -> {
                val listSize = ois.readShort()
                val valueList = ArrayList<Value>()
                for (i in 0 until listSize) {
                    valueList.add(readValue(ois))
                }
                return Value.of(valueList)
            }
            '0' -> return Value.NULL
            'T' -> {
                val time = Instant.ofEpochSecond(ois.readLong(), ois.readLong())
                return time.asValue()
            }
            'S' -> return IOUtils.readString(ois).asValue()
            'D' -> return ois.readDouble().asValue()
            'I' -> return ois.readInt().asValue()
            'L' -> return ois.readLong().asValue()
            'B' -> {
                val intSize = ois.readShort()
                val intBytes = ByteArray(intSize.toInt())
                ois.read(intBytes)
                val scale = ois.readInt()
                val bdc = BigDecimal(BigInteger(intBytes), scale)
                return bdc.asValue()
            }
            'N' -> return Value.of(ois.readObject())
            'X' -> {
                val length = ois.readInt()
                val buffer = ByteArray(length)
                ois.readFully(buffer)
                return BinaryValue(ByteBuffer.wrap(buffer))
            }
            '+' -> return BooleanValue.TRUE
            '-' -> return BooleanValue.FALSE
            '?' -> return ois.readObject() as Value // Read as custom object. Currently reserved
            else -> throw RuntimeException("Wrong value serialization format. Designation $type is unexpected")
        }
    }

    /**
     * Build a meta provider from given general provider
     *
     * @param provider
     * @return
     */
    @JvmStatic
    fun Provider.asValueProvider(): ValueProvider {
        return this as? ValueProvider ?: object : ValueProvider {
            override fun optValue(path: String): Optional<Value> {
                return this@asValueProvider.provide(Path.of(path, ValueProvider.VALUE_TARGET)).map<Value> { Value::class.java.cast(it) }
            }
        }
    }

    private class NumberComparator : Comparator<Number>, Serializable {

        override fun compare(x: Number, y: Number): Int {
            val d1 = x.toDouble()
            val d2 = y.toDouble()
            return if ((d1 != 0.0 || d2 != 0.0) && Math.abs(d1 - d2) / Math.max(d1, d2) < RELATIVE_NUMERIC_PRECISION) {
                0
            } else if (isSpecial(x) || isSpecial(y)) {
                java.lang.Double.compare(d1, d2)
            } else {
                toBigDecimal(x).compareTo(toBigDecimal(y))
            }
        }

        companion object {
            private const val RELATIVE_NUMERIC_PRECISION = 1e-5

            private fun isSpecial(x: Number): Boolean {
                val specialDouble = x is Double && (java.lang.Double.isNaN(x) || java.lang.Double.isInfinite(x))
                val specialFloat = x is Float && (java.lang.Float.isNaN(x) || java.lang.Float.isInfinite(x))
                return specialDouble || specialFloat
            }

            private fun toBigDecimal(number: Number): BigDecimal {
                if (number is BigDecimal) {
                    return number
                }
                if (number is BigInteger) {
                    return BigDecimal(number)
                }
                if (number is Byte || number is Short
                        || number is Int || number is Long) {
                    return BigDecimal(number.toLong())
                }
                if (number is Float || number is Double) {
                    return BigDecimal(number.toDouble())
                }

                try {
                    return BigDecimal(number.toString())
                } catch (e: NumberFormatException) {
                    throw RuntimeException("The given number (\"" + number
                            + "\" of class " + number.javaClass.name
                            + ") does not have a parsable string representation", e)
                }

            }
        }
    }

}

fun ByteBuffer.getValue(): Value{
    val type = get().toChar()
    when (type) {
        '*' -> {
            val listSize = getShort()
            val valueList = ArrayList<Value>()
            for (i in 0 until listSize) {
                valueList.add(getValue())
            }
            return Value.of(valueList)
        }
        '0' -> return Value.NULL
        'T' -> {
            val time = Instant.ofEpochSecond(getLong(), getLong())
            return time.asValue()
        }
        'S' -> {
            val length = getInt()
            val buffer = ByteArray(length)
            get(buffer)
            return String(buffer, Charsets.UTF_8).asValue()
        }
        'D' -> return getDouble().asValue()
        'I' -> return getInt().asValue()
        'L' -> return getLong().asValue()
        'B' -> {
            val intSize = getShort()
            val intBytes = ByteArray(intSize.toInt())
            get()
            val scale = getInt()
            val bdc = BigDecimal(BigInteger(intBytes), scale)
            return bdc.asValue()
        }
        'X' -> {
            val length = getInt()
            val buffer = ByteArray(length)
            get(buffer)
            return BinaryValue(ByteBuffer.wrap(buffer))
        }
        //'N' -> TODO custom number serialization
        '+' -> return BooleanValue.TRUE
        '-' -> return BooleanValue.FALSE
        else -> throw RuntimeException("Wrong value serialization format. Designation $type is unexpected")
    }
}

fun ByteBuffer.putValue(value: Value){
    if (value.isList) {
        put('*'.toByte()) // List designation
        if(value.list.size> Short.MAX_VALUE){
            throw RuntimeException("The array values of size more than ${Short.MAX_VALUE} could not be serialized")
        }
        putShort(value.list.size.toShort())
        value.list.forEach{putValue(it)}
    } else {
        when (value.type) {
            ValueType.NULL -> put('0'.toByte()) // null
            ValueType.TIME -> {
                put('T'.toByte())//Instant
                putLong(value.time.epochSecond)
                putLong(value.time.nano.toLong())
            }
            ValueType.STRING -> {
                put('S'.toByte())//String
                if(value.string.length>Int.MAX_VALUE){
                    throw RuntimeException("The string valuse of size more than ${Int.MAX_VALUE} could not be serialized")
                }
                put(value.string.toByteArray(Charsets.UTF_8))
            }
            ValueType.NUMBER -> {
                val num = value.number
                when (num) {
                    is Double -> {
                        put('D'.toByte()) // double
                       putDouble(num.toDouble())
                    }
                    is Int -> {
                        put('I'.toByte()) // integer
                        putInt(num.toInt())
                    }
                    is Long -> {
                        put('L'.toByte())
                        putLong(num.toLong())
                    }
                    is BigDecimal -> {
                        put('B'.toByte()) // BigDecimal
                        val bigInt = num.unscaledValue().toByteArray()
                        val scale = num.scale()
                        if(bigInt.size> Short.MAX_VALUE){
                            throw RuntimeException("Too large BigDecimal")
                        }
                        putShort(bigInt.size.toShort())
                        put(bigInt)
                        putInt(scale)
                    }
                    else -> {
                        throw RuntimeException("Custom number serialization is not allowed. Yet")
                    }
                }
            }
            ValueType.BOOLEAN -> if (value.boolean) {
                put('+'.toByte()) //true
            } else {
                put('-'.toByte()) // false
            }
            ValueType.BINARY -> {
                put('X'.toByte())
                val binary = value.binary
                putInt(binary.limit())
                put(binary.array())
            }
        }
    }
}
